[RFC] Add support for device extension autoloading

shink opened the pull request:
Fixes #122468

- Load device extensions at the end of `torch/__init__.py`
- Enabled by default, or you can disable it with `TORCH_DEVICE_BACKEND_AUTOLOAD=0`

run test:

```python
python test/run_test.py -i test_autoload_enable
python test/run_test.py -i test_autoload_disable
```

doc:

https://docs-preview.pytorch.org/pytorch/pytorch/127074/miscellaneous_environment_variables.html

co-author:  @jgong5 @bsochack @bkowalskiINTEL @jczaja @FFFrog @hipudding 


cc @albanD

shink committed:
rfc: add device extension autoloading

jgong5 commented:
Thanks for the pull request. Could you please add some tests?

shink commented:
> Thanks for the pull request. Could you please add some tests?

Sure, I'm willing to do this!

bsochack commented:
@shink we pushed a draft PR of a change here: https://github.com/pytorch/pytorch/pull/127228 
A few changes (as in review) + unit tests are still required.

jczaja commented:
@shink , @jgong5  This is a nice work that directly reflect proposal (RFC).  We have tried to test it against our device_extension  and there was a crash : 
![image](https://github.com/pytorch/pytorch/assets/15085062/a3d73833-c21a-4bac-adce-9f46df1a7137)

Problem seems to be that torch._dynamo that is having a code checking all symbols exported from torch/__init__.py . 
So to avoid that it would be good to hide changes proposed here inside a function.  This issue may be observed with extensions that are directly importing torch._dynamo  .  Here is other PR implementing mentioned RFC that works for us: #127228 .

shink commented:
@bsochack @jczaja Nice work! Thanks so much!

> @shink , @jgong5 This is a nice work that directly reflect proposal (RFC). We have tried to test it against our device_extension and there was a crash : ![image](https://private-user-images.githubusercontent.com/15085062/334141500-a3d73833-c21a-4bac-adce-9f46df1a7137.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY4NTkyMjMsIm5iZiI6MTcxNjg1ODkyMywicGF0aCI6Ii8xNTA4NTA2Mi8zMzQxNDE1MDAtYTNkNzM4MzMtYzIxYS00YmFjLWFkY2UtOWY0NmRmMWE3MTM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI4VDAxMTUyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjYjc4MzgyMmU3YjRmMzQ5NzhmNWM0ZmQ5NzA1ODA0NjY4ZTRmYmUwMTNlMzQ1MjRkMDY0NzU2NjhjY2I0YjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6wQP0i5S9aP6IhfGGSyFUxU4SgqhL13cymxgoGdJnYY)
> 
> Problem seems to be that torch._dynamo is having a code checking all symbols exported from torch/**init**.py . So to avoid that it would be good to hide changes proposed here inside a function. This issue may be observed with extensions that are directly importing torch._dynamo . Here is other PR implementing mentioned RFC that works for us: #127228 .

Yes, we alse met this crash.

shink committed:
test: add ut for autoloading

shink commented:
@jgong5 @bsochack @bkowalskiINTEL @FFFrog @hipudding Please have a look at this unit test. See https://github.com/pytorch/pytorch/pull/127074/commits/b0e08d39ec696e71d989ef4abdc8089599366b58

shink commented on test/autoload/device_backend/device_backend.egg-info/entry_points.txt:
```diff
@@ -0,0 +1,2 @@
+[torch.backends]
+device_backend = backend_pkg:autoload
```
in fact, this file should be ignored, and the reason it is here is to declear the entry point of the backend package

shink committed:
style: fix lint issues

FFFrog commented on torch/__init__.py:
```diff
@@ -2077,3 +2077,27 @@ def _constrain_as_size(symbol, min: Optional[builtins.int] = None, max: Optional
+        try:
+            # just load the plugin without calling
+            backend.load()
+        except Exception:
```
I think it would be better to show failure information when an exception is encountered.

FFFrog commented on torch/__init__.py:
```diff
@@ -2077,3 +2077,27 @@ def _constrain_as_size(symbol, min: Optional[builtins.int] = None, max: Optional
+
+
+def is_device_backend_autoload_enabled() -> bool:
+    var = os.getenv("TORCH_DISABLE_DEVICE_BACKEND_AUTOLOAD")
```
`getenv` with default value shoule be better.

FFFrog commented on test/autoload/test_autoload.py:
```diff
@@ -0,0 +1,32 @@
+
+class TestAutoload(TestCase):
+    def test_load_plugins(self):
+        device_backend_path = os.path.abspath(
```
It would be better to think of this path operation as context.

bsochack commented:
> @jgong5 @bsochack @bkowalskiINTEL @FFFrog @hipudding Please have a look at this unit test. See [b0e08d3](https://github.com/pytorch/pytorch/commit/b0e08d39ec696e71d989ef4abdc8089599366b58)

Please add @bkowalskiINTEL and @jczaja as co-authors. Lets also continue all addressing code review here.

shink commented:
> > @jgong5 @bsochack @bkowalskiINTEL @FFFrog @hipudding Please have a look at this unit test. See [b0e08d3](https://github.com/pytorch/pytorch/commit/b0e08d39ec696e71d989ef4abdc8089599366b58)
> 
> Please add @bkowalskiINTEL and @jczaja as co-authors. Lets also continue all addressing code review here.

Sure, thanks so much.

shink committed:
test: update code

shink committed:
chore

shink committed:
test: update code

jgong5 commented on test/autoload/test_autoload.py:
```diff
@@ -0,0 +1,38 @@
+    def test_autoload(self):
+        # after importing the extension, the value of this environment variable should be true
+        torch.import_device_backends()
+        value = os.getenv("IS_CUSTOM_DEVICE_BACKEND_IMPORTED", "false")
```
Besides checking the environment variable, are you going to add test to autoload an extension (maybe a mock one)?

shink replied:
Thanks for your review. Please have a look at the code I just updated.

shink committed:
test: update code

shink committed:
docs: add README

albanD commented on torch/__init__.py:
```diff
@@ -2077,3 +2077,43 @@ def _constrain_as_size(symbol, min: Optional[builtins.int] = None, max: Optional
+
+def _is_device_backend_autoload_enabled() -> bool:
+    """
+    Whether to enbale autoloading out-of-the-tree device extensions.
```
```suggestion
    Whether autoloading out-of-the-tree device extensions is enabled.
```

albanD commented on torch/__init__.py:
```diff
@@ -2077,3 +2077,43 @@ def _constrain_as_size(symbol, min: Optional[builtins.int] = None, max: Optional
+    """
+    # enabled by default
+    is_enable = os.getenv("TORCH_DEVICE_BACKEND_AUTOLOAD", "1")
+    return is_enable.strip().lower() in ("1", "true", "yes", "on", "y")
```
```suggestion
    return is_enable.strip().lower() in {"1", "true", "yes", "on", "y"}
```

albanD commented on torch/__init__.py:
```diff
@@ -2077,3 +2077,43 @@ def _constrain_as_size(symbol, min: Optional[builtins.int] = None, max: Optional
+            # Just load the extension without calling
+            backend_extension.load()
+        except Exception as err:
+            print(f"Failed to load the backend extension: {backend_extension.name}")
```
Why silence this failure rather than letting it stop the user code? 
```suggestion
            raise RuntimeError(f"Failed to load the backend extension: {backend_extension.name}. You can disable extension auto-loading with TORCH_DEVICE_BACKEND_AUTOLOAD=0.") from err
```

albanD commented on test/autoload/device_backend/README.md:
````diff
@@ -0,0 +1,13 @@
+```bash
+test/autoload/device_backend
+├── README.md
+├── backend_pkg               # The backend package
````
Would it be possible to reuse the existing modules we already have for testing in `test/cpp_extension/setup.py` to test this without having to create a brand new extension?

shink replied:
Thank you so much for your review. I'm testing this idea now.

shink replied:
Could you please have a look at this commit https://github.com/pytorch/pytorch/pull/127074/commits/1e7cf76cb4fc7b94aae267c70ba37d75a8b93282

albanD commented on test/autoload/test_autoload.py:
```diff
@@ -0,0 +1,47 @@
+        value = os.getenv("IS_CUSTOM_DEVICE_BACKEND_IMPORTED", "false")
+        self.assertEqual(value, "true")
+
+        # Test the function defined in backend_pkg/__init__.py
+        import backend_pkg
+        self.assertTrue(hasattr(backend_pkg, "apply_patch"))
+        self.assertEqual(backend_pkg.apply_patch(), "success")
```
I'm not sure what this is testing?

shink replied:
it's testing a function defined in the backend extension

shink committed:
Update torch/__init__.py

Co-authored-by: albanD <desmaison.alban@gmail.com>

shink committed:
Update torch/__init__.py

Co-authored-by: albanD <desmaison.alban@gmail.com>

shink committed:
Update torch/__init__.py

let the failure stop the user code

Co-authored-by: albanD <desmaison.alban@gmail.com>

shink committed:
reuse the cpp_extensions module

shink committed:
fix: fix issues and use cpp_extensions

shink committed:
fix: update test

shink committed:
fix: change test case location

jgong5 commented:
@shink Seems there is still lint errors. You may repro with "lintrunner -a" locally.

jgong5 commented on test/cpp_extensions/torch_test_cpp_extension/__init__.py:
```diff
@@ -0,0 +1,14 @@
+
+import os
+
+# When importing this package, set this environment variable to true
+os.environ["IS_CUSTOM_DEVICE_BACKEND_IMPORTED"] = "true"
+
+
+def _autoload():
+    # Do nothing in this entry point
+    pass
```
Should we do things inside this entrypoint instead to make sure it is called?

jgong5 commented on test/test_cpp_extensions_aot.py:
```diff
@@ -367,5 +367,13 @@ def f(a: bool, b: bool):
 
 
+class TestDeviceBackendAutoload(common.TestCase):
+    def test_autoload(self):
```
Not sure if I understand it correctly but when the cpp extension is explicitly imported from this test file, not autoloaded implicitly. Is that correct?

shink replied:
The reason for putting this test case here is that I want to reuse the existing modules in `test/cpp_extension/setup.py`.

shink committed:
fix: call the entrypoint

shink committed:
test: add test case

shink committed:
test: update

shink committed:
revert

shink committed:
fix: fix compile

shink committed:
test: add test_autoload_enable and test_autoload_disable

shink committed:
test: add test_autoload_enable and test_autoload_disable (part 2)

shink committed:
test: add test_autoload_enable and test_autoload_disable (part 3)

shink commented:
@jgong5 Please take a look, thanks so much.

jgong5 commented on test/test_autoload.py:
```diff
@@ -0,0 +1,24 @@
+        is_imported = os.getenv("IS_CUSTOM_DEVICE_BACKEND_IMPORTED", "False")
+
+        # Both values should be equal
+        if switch:
+            self.assertEqual(is_imported, switch)
```
```suggestion
        if switch:
            self.assertEqual(is_imported, switch)
        else:
            self.assertTrue(is_imported)
```

jgong5 commented:
@pytorch rebase

jgong5 commented:
@pytorchbot rebase

Commit:
Update test/test_autoload.py

jgong5 commented:
@albanD The PR has been updated according to your comment. Could you please take a look again? We are proposing to make it as part of PyTorch 2.4 prototype features so would like to get it landed before the branch cut due 6/10. Thanks!

jgong5 commented:
@albanD Does the PR look good to you?

bsochack commented:
@albanD could you please take a look at this PR?

albanD commented:
Sorry I've been quite busy, will look by the end of the week.

albanD commented on test/test_autoload.py:
```diff
@@ -0,0 +1,26 @@
+
+class TestDeviceBackendAutoload(TestCase):
+    def test_autoload(self):
+        switch = os.getenv("TORCH_DEVICE_BACKEND_AUTOLOAD", None)
```
```suggestion
        switch = os.getenv("TORCH_DEVICE_BACKEND_AUTOLOAD", False)
```
If you do this, you won't need the `if switch` below right?

shink replied:
Thanks for your review, I will test this change.

albanD commented on test/run_test.py:
```diff
@@ -717,6 +717,51 @@ def test_cpp_extensions_aot_no_ninja(test_module, test_directory, options):
+    return _test_autoload(test_directory, options, enable=False)
+
+
+def _test_autoload(test_directory, options, enable=True):
```
FYI @clee2000 for change to run_test.py

albanD commented on torch/__init__.py:
```diff
@@ -2112,3 +2112,51 @@ def _constrain_as_size(symbol, min: Optional[builtins.int] = None, max: Optional
+    """
+    # enabled by default
+    is_enable = os.getenv("TORCH_DEVICE_BACKEND_AUTOLOAD", "1")
+    return is_enable.strip().lower() in {"1", "true", "yes", "on", "y"}
```
nit: we usually do "0" and "1" only for these for simplicity.

albanD reviewed (commented):
Small things only!

shink committed:
Merge branch 'main' into feat/autoload

shink committed:
improve test code

bsochack commented:
@albanD is this PR ready to merge?

albanD commented on test/test_autoload.py:
```diff
@@ -0,0 +1,23 @@
+# Owner(s): ["module: unknown"]
```
```suggestion
# Owner(s): ["module: PrivateUse1"]
```

albanD commented on test/test_autoload.py:
```diff
@@ -0,0 +1,23 @@
+
+import os
+
+import torch  # noqa: F401
```
Why do you need this? The import of torch.testing below will do this by side effect.

albanD commented on torch/__init__.py:
```diff
@@ -2473,3 +2473,50 @@ def _constrain_as_size(
+    """
+    Whether autoloading out-of-the-tree device extensions is enabled.
+    The switch depends on the value of the environment variable
+    `TORCH_DEVICE_BACKEND_AUTOLOAD`.
```
@drisspg which .rst file should we add this new env variable to for proper documentation?

drisspg replied:
The main env var RST is here:https://github.com/pytorch/pytorch/blob/main/docs/source/torch_environment_variables.rst

and they branch out to others

shink committed:
Update test/test_autoload.py [skip ci]

Co-authored-by: albanD <desmaison.alban@gmail.com>

shink committed:
remove redundant import

shink committed:
docs: add env var rst

shink committed:
docs: fix format

bsochack commented:
@albanD any comments to this PR?

albanD commented on docs/source/torch_environment_variables.rst:
```diff
@@ -26,3 +26,4 @@ If you find anything in this documentation that is missing, incorrect, or could
    miscellaneous_environment_variables
    logging
    torch_nccl_environment_variables
+   privateuse1_environment_variables
```
Could you please move this to the miscellaneous section? I do expect all users to be interested in this env variable, not just privateuse1 devs.
https://github.com/pytorch/pytorch/blob/34e94c507ab508ab6dea61373dc93450a9618db8/docs/source/miscellaneous_environment_variables.rst#L13

shink replied:
@albanD Sure, it has been moved. Please take a look again. Thanks so much for your patient review.

albanD reviewed (approved):
Small nit on moving the doc to another file, but SGTM otherwise!

shink committed:
docs: move the env to miscellaneous_environment_variables.rst

jgong5 commented:
@pytorchbot merge

jgong5 commented:
@pytorchbot merge
